File: 07-Weighted-Graph/Code-Python/edge.py
"""
带权图的边
"""


class Edge:
    def __init__(self, a, b, weight):
        self.a = a
        self.b = b
        self.weight = weight

    def get_v(self):
        """返回第一个顶点"""
        return self.a

    def get_w(self):
        """返回第二个顶点"""
        return self.b

    def get_weight(self):
        """返回权值"""
        return self.weight

    def other(self, x):
        """给定一个顶点, 返回另一个顶点"""
        assert x == self.a or x == self.b
        return self.a if x == self.b else self.b

    def __lt__(self, other):
        return self.weight < other.weight

    def __gt__(self, other):
        return self.weight > other.weight

    def __str__(self):
        return '' + str(self.a) + '-' + str(self.b) + ': ' + str(self.weight)

File: 07-Weighted-Graph/Code-Java/Edge.java
package weightedGraph;

/**
 * 带权图的边
 */
public class Edge<Weight extends Number & Comparable> implements Comparable<Edge> {
    private int a, b;  //边的两个端点
    private Weight weight;  //边的权值

    public Edge(int a, int b, Weight weight) {
        this.a = a;
        this.b = b;
        this.weight = weight;
    }

    public Edge(Edge<Weight> e) {
        this.a = e.a;
        this.b = e.b;
        this.weight = e.weight;
    }

    // 返回第一个顶点
    public int getV() {
        return a;
    }

    // 返回第二个顶点
    public int getW() {
        return b;
    }

    // 返回权值
    public Weight getWeight() {
        return weight;
    }

    // 给定一个顶点, 返回另一个顶点
    public int other(int x) {
        assert x == a || x == b;
        return x == a ? b : a;
    }

    // 输出边的信息
    @Override
    public String toString() {
        return "" + a + "-" + b + ": " + weight;
    }

    // 边之间的比较
    @Override
    public int compareTo(Edge o) {
        if (weight.compareTo(o.getWeight()) < 0)
            return -1;
        else if (weight.compareTo(o.getWeight()) > 0)
            return +1;
        else return 0;
    }
}
